cartographer/io/fake_file_writer.cc

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/io/fake_file_writer.h"
+
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace io {
+
+FakeFileWriter::FakeFileWriter(const std::string& filename,
+                               std::shared_ptr<std::string> content)
+    : is_closed_(false), content_(content), filename_(filename) {
+  CHECK(content);
+}
+
+bool FakeFileWriter::Write(const char* const data, const size_t len) {
+  if (is_closed_) return false;
+  content_->append(data, len);
+  return true;
+}
+
+bool FakeFileWriter::Close() {
+  if (is_closed_) return false;
+  is_closed_ = true;
+  return true;
+}
+
+bool FakeFileWriter::WriteHeader(const char* const data, const size_t len) {
+  if (is_closed_) return false;
+  if (content_->size() == 0 || content_->size() < len) {
+    *content_ = "";
+    return Write(data, len);
+  }
+  content_->replace(0, len, data);
```

Comment:
This is not what we want. The first `len` characters should be replaced by `len` bytes from `data`. This code assumes data is a null terminated string. Add a test that fails with the current code, then replace? How about `std::copy(data, data + len, content_.begin());` or so?

Reply:
You're right. I changed the content member from std::string to std::vector<char> to avoid any bugs resulting from using string methods.